keps/sig-node/4960-container-stop-signals/README.md

```diff
@@ -0,0 +1,485 @@
+
+Having stop signal as a first class citizen in the Pod's container specification would make it easier for users to set custom stop signals for their containers across all types of workloads.
+
+### Goals
+
+- Add a new Stop lifecycle handler to container lifecycle which can be configured with a Signal option
+- Update the CRI API to pass down the stop signal to the container runtime via ContainerConfig
+- Update the implementation of the StopContainer method in container runtimes to use the container’s stop signal (if defined) to kill containers
+- Add support to show the effective stop signal of containers in the container status field in the pod status
+
+### Non-Goals
+
+- Change any existing behaviour with how stop signals work when defined in the container image
+
+## Proposal
+
+### API
+
+A new Stop lifecycle handler will be added to container lifecycle. The Stop lifecycle event can be configured with a Signal option, which is of type `Signal`. This new `Signal` type can take a string value, and can be used to define a stop signal for containers when creating Pods. `Signal` will hold string values which can be mapped to Go's syscall.Signal. For example, [list of signals supported in Linux environments by moby](https://github.com/containerd/containerd/blob/main/vendor/github.com/moby/sys/signal/signal_linux.go). If the user doesn't define a particular stop signal, the behaviour would default to what it is today and fallback to the stop signal defined in the container image or use the default stop signal of the container runtime (SIGTERM in case of containerd, CRI-O).
+
+```go
+// pkg/apis/core/types.go
+type Signal string //parseable into Go's syscall.Signal
+
+type LifecycleHandler struct {
+  // ...
+	// +optional
+	Signal *Signal
+}
+
+type Lifecycle struct {
+  // ...
+	// +optional
+	Stop *LifecycleHandler
+}
+```
+
+Users will be able to define custom stop signals for their containers like so:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: nginx
+spec:
+  containers:
+  - name: nginx
+    image: nginx:1.14.2
+    lifecycle:
+      stop:
+        signal: SIGUSR1
+```
+
+The stop signal would also be shown in the containers' status. The value of the stop signal shown in the status can be from the spec, if a stop cycle is defined in the spec, else it will be the effective stop signal which is used by the container runtime to kill your container. This can either be read from the container image or will be the default stop signal of the container runtime. Users will be able to see a container's stop signal in its status even if they're not using a custom stop signal from the spec.
+
+```yaml
+status:
+  containerStatuses:
+  - containerID: containerd://19d9bb24f5d6633dddf8f97d0b2aed1158ceb1030440082f3f3dbea8ce4d2be6
+    image: nginx:1.14.2
+    lastState: {}
+    lifecyle:
+      stop:
```

Comment:
Options:

```
    lifecycle:
      stop:
        signal: SIGUSR1
```

or

```
    lifecycle:
      stopSignal: SIGUSR1
```

I have a slight lean toward the second beause a signal is always sent.  If we added  `lifecycle.stop.*` that would be IN ADDITION TO, not INSTEAD OF the signal.

thoughts?

Reply:
I'm not super concerned about that. It was opportunistic.  Let's go with the second option as the plan? Just a string field, no new struct

Reply:
Yep, I've updated the KEP in https://github.com/kubernetes/enhancements/pull/5122/commits/760c51d1e9406bea3643764b8fc6f9a2637f6a09 to have `lifecycle.stopSignal` instead of `lifecycle.stop.signal`.

One more question: we say that a signal is always sent, but in the case where `pod.spec.containers[].lifecycle.stopSignal` is nil or set to "", we're letting the runtime choose the signal. Would this contradict us saying that "we always send a signal"? To phrase this question in another way, the source of truth for the effective stop signal doesn't need to be the kubelet/container spec right? The effective signal would be present in the status, but this value can come from either the spec or the container image/runtime default.

Reply:
> Would this contradict us saying that "we always send a signal"?

I don't think so.  We are still going to send a signal, the pod just didn't have an opinion on which exact signal.

> the source of truth for the effective stop signal doesn't need to be the kubelet/container spec right?

Correct.  The true value of the status field is to express what happened as a result of spec.  If you did not specify it, you got SIGTERM.  It's explicit, and not implied.

Reply:
Ah that makes sense, thanks a lot!